bn-apps/cordapp-updates-distribution/corda-updates-app/src/main/kotlin/com/r3/businessnetworks/cordaupdates/app/bno/DatabaseService.kt

```diff
@@ -1,130 +1,136 @@
-        val insertQuery = """
-                insert into $CORDAPP_VERSION_INFO_TABLE ($PARTY, $CORDAPP_GROUP, $CORDAPP_NAME, $CORDAPP_VERSION, $LAST_UPDATED)
-                values (?, ?, ?, ?, ?)
-                """
-        val insertStatement : PreparedStatement = session.prepareStatement(insertQuery)
-        insertStatement.setString(1, party.name.toString())
-        insertStatement.setString(2, cordappVersionInfo.group)
-        insertStatement.setString(3, cordappVersionInfo.name)
-        insertStatement.setString(4, cordappVersionInfo.version)
-        insertStatement.setLong(5, System.currentTimeMillis())
-        insertStatement.executeUpdate()
+    private fun insertCordappVersionInfo(party : Party, cordappVersionInfo : CordappVersionInfo) {
+        serviceHub.withEntityManager {
+            persist(PersistentCordappVersionInfo.from(party, cordappVersionInfo))
+        }
     }
 
     /**
      * Get information about all reported cordapps versions
      */
     fun getCordappVersionInfos() : Map<String, List<CordappVersionInfo>> {
-        val nativeQuery = """
-                select * from $CORDAPP_VERSION_INFO_TABLE
-            """
-        val session : Connection = serviceHub.jdbcSession()
-        val resultSet : ResultSet = session.prepareStatement(nativeQuery).use { it.executeQuery()!! }
         val results = mapOf<String, MutableList<CordappVersionInfo>>()
-        while (resultSet.next()) {
-            val party : String = resultSet.getString(PARTY)!!
-            val versions : MutableList<CordappVersionInfo> = results.getOrElse(party) { mutableListOf() }
-            versions.add(
-                    CordappVersionInfo(resultSet.getString(CORDAPP_GROUP)!!,
-                            resultSet.getString(CORDAPP_NAME)!!,
-                            resultSet.getString(CORDAPP_VERSION)!!,
-                            resultSet.getLong(LAST_UPDATED)))
+
+        serviceHub.withEntityManager {
+            createQuery("from PersistentCordappVersionInfo")
+                    .resultList
+                    .map {
+                        val persistentCordappVersionInfo = it as PersistentCordappVersionInfo
+                        val versions : MutableList<CordappVersionInfo> = results.getOrElse(persistentCordappVersionInfo.party.toString()) { mutableListOf() }
+                        versions.add(it.toCordappVersionInfo())
+                    }
         }
+
         return results
     }
 
     /**
      * Get information about reported cordapps versions for the provided party
      */
     fun getCordappVersionInfos(party : Party) : List<CordappVersionInfo> {
-        val nativeQuery = """
-                select * from $CORDAPP_VERSION_INFO_TABLE where $PARTY='${party.name}'
+        return serviceHub.withEntityManager {
+            val nativeQuery = """
+                from PersistentCordappVersionInfo where party=:party
```

Comment:
I'd avoid native queries if in favour of `CriteriaBuilder` for portability. For simple statements, it shouldn't make a difference but for more complex ones it might.

Reply:
Agreed, again this is HQL, which gets converted to a db-engine specific syntax. Let me know if you still think that it needs replacing - would be happy to do that.

Reply:
OK my bad, didn't know that. I guess the programmatic way feels more hibernate-y which tripped me up, but never mind then.